{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fed3dafd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO, SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import os\n",
    "import glob\n",
    "import pandas as pd\n",
    "from Bio.SeqRecord import SeqRecord\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "7da72130",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cross_check(fasta_path, region):\n",
    "    SPIKE_COORDS = (21563, 25321)\n",
    "    RBD_COORDS = (22553, 23155)\n",
    "    (start, end) = SPIKE_COORDS if region == \"spike\" else RBD_COORDS\n",
    "\n",
    "    rec = SeqIO.read(fasta_path, format=\"fasta\")\n",
    "\n",
    "    wt = []\n",
    "    aa_pos_ls = []\n",
    "    coords = []\n",
    "    aa_pos = 1 if region == \"spike\" else 331 # 1-based idx\n",
    "\n",
    "    for pos in range(start-1, end, 3): # 0-based idx\n",
    "        try:\n",
    "            ref_seq = str(rec.seq)\n",
    "            aa = str(Seq(ref_seq[pos: pos+3]).translate())\n",
    "\n",
    "            if aa  == \"*\" or aa == \"-\":\n",
    "                aa_pos += 1\n",
    "                continue\n",
    "            wt.append(aa)\n",
    "            aa_pos_ls.append(aa_pos)\n",
    "            coords.append((int(pos), int(pos+3))) # 0-based idx\n",
    "            aa_pos += 1\n",
    "        except:\n",
    "            aa_pos += 1\n",
    "            continue\n",
    "        \n",
    "    df = pd.DataFrame({'wildtype': wt, 'site': aa_pos_ls, \"coords\": coords}) if region == \"spike\" else pd.DataFrame({'wt': wt, 'pos': aa_pos_ls, \"coords\": coords})\n",
    "\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e95bd5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wildtype</th>\n",
       "      <th>site</th>\n",
       "      <th>coords</th>\n",
       "      <th>mutant</th>\n",
       "      <th>spike mediated entry</th>\n",
       "      <th>ACE2 binding</th>\n",
       "      <th>sequential_site</th>\n",
       "      <th>region</th>\n",
       "      <th>_merge</th>\n",
       "      <th>codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>N</td>\n",
       "      <td>422</td>\n",
       "      <td>(22825, 22828)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>aac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>N</td>\n",
       "      <td>437</td>\n",
       "      <td>(22870, 22873)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>aac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520</th>\n",
       "      <td>V</td>\n",
       "      <td>524</td>\n",
       "      <td>(23131, 23134)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>gtg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>559</th>\n",
       "      <td>Q</td>\n",
       "      <td>563</td>\n",
       "      <td>(23248, 23251)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>caa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>I</td>\n",
       "      <td>720</td>\n",
       "      <td>(23719, 23722)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>atc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>S</td>\n",
       "      <td>816</td>\n",
       "      <td>(24007, 24010)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>tcc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>814</th>\n",
       "      <td>I</td>\n",
       "      <td>818</td>\n",
       "      <td>(24013, 24016)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>att</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970</th>\n",
       "      <td>S</td>\n",
       "      <td>974</td>\n",
       "      <td>(24481, 24484)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>tcc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    wildtype  site          coords mutant  spike mediated entry  ACE2 binding  \\\n",
       "418        N   422  (22825, 22828)    NaN                   NaN           NaN   \n",
       "433        N   437  (22870, 22873)    NaN                   NaN           NaN   \n",
       "520        V   524  (23131, 23134)    NaN                   NaN           NaN   \n",
       "559        Q   563  (23248, 23251)    NaN                   NaN           NaN   \n",
       "716        I   720  (23719, 23722)    NaN                   NaN           NaN   \n",
       "812        S   816  (24007, 24010)    NaN                   NaN           NaN   \n",
       "814        I   818  (24013, 24016)    NaN                   NaN           NaN   \n",
       "970        S   974  (24481, 24484)    NaN                   NaN           NaN   \n",
       "\n",
       "     sequential_site region     _merge codon  \n",
       "418              NaN    NaN  left_only   aac  \n",
       "433              NaN    NaN  left_only   aac  \n",
       "520              NaN    NaN  left_only   gtg  \n",
       "559              NaN    NaN  left_only   caa  \n",
       "716              NaN    NaN  left_only   atc  \n",
       "812              NaN    NaN  left_only   tcc  \n",
       "814              NaN    NaN  left_only   att  \n",
       "970              NaN    NaN  left_only   tcc  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_path = \"/home/yutianc/sars-cov-2-dms-ref/output/NC_045512.2_escape_ba.2_spike.fa\"\n",
    "csv_path = \"/home/yutianc/sars-cov-2-dms-ref/data/ba.2/spike/BA.2_spike_summary.csv\"\n",
    "\n",
    "df = cross_check(fasta_path, region=\"spike\")\n",
    "df_true = pd.read_csv(csv_path)\n",
    "df_true = df_true.drop_duplicates(subset=[\"site\"])\n",
    "ref_seq = str(SeqIO.read(fasta_path, format=\"fasta\").seq)\n",
    "\n",
    "m = pd.merge(df, df_true, how=\"outer\", on=[\"site\", \"wildtype\"], indicator=True)\n",
    "m = m[m[\"_merge\"] != \"both\"]\n",
    "m[\"codon\"] = m[\"coords\"].apply(lambda x: ref_seq[x[0]: x[1]])\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "9742e332",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wildtype</th>\n",
       "      <th>site</th>\n",
       "      <th>coords</th>\n",
       "      <th>mutant</th>\n",
       "      <th>human sera escape</th>\n",
       "      <th>spike mediated entry</th>\n",
       "      <th>ACE2 binding</th>\n",
       "      <th>sequential_site</th>\n",
       "      <th>region</th>\n",
       "      <th>_merge</th>\n",
       "      <th>codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>V</td>\n",
       "      <td>143</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.03895</td>\n",
       "      <td>-0.08258</td>\n",
       "      <td>-0.3188</td>\n",
       "      <td>140.0</td>\n",
       "      <td>NTD</td>\n",
       "      <td>right_only</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>N</td>\n",
       "      <td>422</td>\n",
       "      <td>(22825, 22828)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>aac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>C</td>\n",
       "      <td>480</td>\n",
       "      <td>(22999, 23002)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>tgt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501</th>\n",
       "      <td>Q</td>\n",
       "      <td>506</td>\n",
       "      <td>(23077, 23080)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>caa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>526</th>\n",
       "      <td>T</td>\n",
       "      <td>531</td>\n",
       "      <td>(23152, 23155)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>acc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>558</th>\n",
       "      <td>Q</td>\n",
       "      <td>563</td>\n",
       "      <td>(23248, 23251)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>caa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>P</td>\n",
       "      <td>579</td>\n",
       "      <td>(23296, 23299)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>cca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715</th>\n",
       "      <td>I</td>\n",
       "      <td>720</td>\n",
       "      <td>(23719, 23722)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>atc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>E</td>\n",
       "      <td>725</td>\n",
       "      <td>(23734, 23737)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>gag</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>744</th>\n",
       "      <td>C</td>\n",
       "      <td>749</td>\n",
       "      <td>(23806, 23809)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>tgt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>813</th>\n",
       "      <td>I</td>\n",
       "      <td>818</td>\n",
       "      <td>(24013, 24016)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>att</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914</th>\n",
       "      <td>N</td>\n",
       "      <td>919</td>\n",
       "      <td>(24316, 24319)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>aac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>965</th>\n",
       "      <td>F</td>\n",
       "      <td>970</td>\n",
       "      <td>(24469, 24472)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>ttt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>969</th>\n",
       "      <td>S</td>\n",
       "      <td>974</td>\n",
       "      <td>(24481, 24484)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>left_only</td>\n",
       "      <td>tcc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    wildtype  site          coords mutant  human sera escape  \\\n",
       "139        V   143             NaN      -           -0.03895   \n",
       "417        N   422  (22825, 22828)    NaN                NaN   \n",
       "475        C   480  (22999, 23002)    NaN                NaN   \n",
       "501        Q   506  (23077, 23080)    NaN                NaN   \n",
       "526        T   531  (23152, 23155)    NaN                NaN   \n",
       "558        Q   563  (23248, 23251)    NaN                NaN   \n",
       "574        P   579  (23296, 23299)    NaN                NaN   \n",
       "715        I   720  (23719, 23722)    NaN                NaN   \n",
       "720        E   725  (23734, 23737)    NaN                NaN   \n",
       "744        C   749  (23806, 23809)    NaN                NaN   \n",
       "813        I   818  (24013, 24016)    NaN                NaN   \n",
       "914        N   919  (24316, 24319)    NaN                NaN   \n",
       "965        F   970  (24469, 24472)    NaN                NaN   \n",
       "969        S   974  (24481, 24484)    NaN                NaN   \n",
       "\n",
       "     spike mediated entry  ACE2 binding  sequential_site region      _merge  \\\n",
       "139              -0.08258       -0.3188            140.0    NTD  right_only   \n",
       "417                   NaN           NaN              NaN    NaN   left_only   \n",
       "475                   NaN           NaN              NaN    NaN   left_only   \n",
       "501                   NaN           NaN              NaN    NaN   left_only   \n",
       "526                   NaN           NaN              NaN    NaN   left_only   \n",
       "558                   NaN           NaN              NaN    NaN   left_only   \n",
       "574                   NaN           NaN              NaN    NaN   left_only   \n",
       "715                   NaN           NaN              NaN    NaN   left_only   \n",
       "720                   NaN           NaN              NaN    NaN   left_only   \n",
       "744                   NaN           NaN              NaN    NaN   left_only   \n",
       "813                   NaN           NaN              NaN    NaN   left_only   \n",
       "914                   NaN           NaN              NaN    NaN   left_only   \n",
       "965                   NaN           NaN              NaN    NaN   left_only   \n",
       "969                   NaN           NaN              NaN    NaN   left_only   \n",
       "\n",
       "    codon  \n",
       "139  None  \n",
       "417   aac  \n",
       "475   tgt  \n",
       "501   caa  \n",
       "526   acc  \n",
       "558   caa  \n",
       "574   cca  \n",
       "715   atc  \n",
       "720   gag  \n",
       "744   tgt  \n",
       "813   att  \n",
       "914   aac  \n",
       "965   ttt  \n",
       "969   tcc  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_path = \"/home/yutianc/sars-cov-2-dms-ref/output/NC_045512.2_escape_xbb.1.5_spike.fa\"\n",
    "csv_path = \"/home/yutianc/bjorn_rep/data/sc2/muninn_inputs/dms/XBB.1.5_spike_DMS.csv\"\n",
    "\n",
    "df = cross_check(fasta_path, region=\"spike\")\n",
    "df_true = pd.read_csv(csv_path)\n",
    "df_true = df_true.drop_duplicates(subset=[\"site\"])\n",
    "ref_seq = str(SeqIO.read(fasta_path, format=\"fasta\").seq)\n",
    "\n",
    "m = pd.merge(df, df_true, how=\"outer\", on=[\"site\", \"wildtype\"], indicator=True)\n",
    "m = m[m[\"_merge\"] != \"both\"]\n",
    "m[\"codon\"] = m[\"coords\"].apply(lambda x: ref_seq[x[0]: x[1]] if pd.notna(x) else None)\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e9549b8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "ffcfb60f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt</th>\n",
       "      <th>pos</th>\n",
       "      <th>coords</th>\n",
       "      <th>mut1</th>\n",
       "      <th>num_aa</th>\n",
       "      <th>_merge</th>\n",
       "      <th>codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [wt, pos, coords, mut1, num_aa, _merge, codon]\n",
       "Index: []"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_path = \"/home/yutianc/bjorn_rep/data/sc2/NC_045512.2_escape_BA.1_rbd.fa\"\n",
    "csv_path = \"/home/yutianc/sars-cov-2-dms-ref/data/ba.1/mut1_BA.1_Omicron_baseline_EPI_ISL_10000028.csv\"\n",
    "\n",
    "df = cross_check(fasta_path, region=\"rbd\")\n",
    "df_true = pd.read_csv(csv_path)\n",
    "df_true = df_true.drop_duplicates(subset=[\"pos\"])\n",
    "ref_seq = str(SeqIO.read(fasta_path, format=\"fasta\").seq)\n",
    "\n",
    "m = pd.merge(df, df_true, how=\"outer\", on=[\"pos\", \"wt\"], indicator=True)\n",
    "m = m[m[\"_merge\"] != \"both\"]\n",
    "m[\"codon\"] = m[\"coords\"].apply(lambda x: ref_seq[x[0]: x[1]])\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "c0ae54e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt</th>\n",
       "      <th>pos</th>\n",
       "      <th>coords</th>\n",
       "      <th>mut1</th>\n",
       "      <th>num_aa</th>\n",
       "      <th>_merge</th>\n",
       "      <th>codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [wt, pos, coords, mut1, num_aa, _merge, codon]\n",
       "Index: []"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_path = \"/home/yutianc/bjorn_rep/data/sc2/NC_045512.2_escape_BA.2_rbd.fa\"\n",
    "csv_path = \"/home/yutianc/sars-cov-2-dms-ref/data/ba.2/rbd/mut1_BA.2_Omicron_baseline_EPI_ISL_10000005.csv\"\n",
    "\n",
    "df = cross_check(fasta_path, region=\"rbd\")\n",
    "df_true = pd.read_csv(csv_path)\n",
    "df_true = df_true.drop_duplicates(subset=[\"pos\"])\n",
    "ref_seq = str(SeqIO.read(fasta_path, format=\"fasta\").seq)\n",
    "\n",
    "m = pd.merge(df, df_true, how=\"outer\", on=[\"pos\", \"wt\"], indicator=True)\n",
    "m = m[m[\"_merge\"] != \"both\"]\n",
    "m[\"codon\"] = m[\"coords\"].apply(lambda x: ref_seq[x[0]: x[1]])\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "a7aeaa05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt</th>\n",
       "      <th>pos</th>\n",
       "      <th>coords</th>\n",
       "      <th>mut1</th>\n",
       "      <th>num_aa</th>\n",
       "      <th>_merge</th>\n",
       "      <th>codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [wt, pos, coords, mut1, num_aa, _merge, codon]\n",
       "Index: []"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_path = \"/home/yutianc/bjorn_rep/data/sc2/NC_045512.2_escape_BA.2.75_rbd.fa\"\n",
    "csv_path = \"/home/yutianc/sars-cov-2-dms-ref/data/ba.2.75/mut1_BA.2.75_EPI_ISL_13302209.csv\"\n",
    "\n",
    "df = cross_check(fasta_path, region=\"rbd\")\n",
    "df_true = pd.read_csv(csv_path)\n",
    "df_true = df_true.drop_duplicates(subset=[\"pos\"])\n",
    "ref_seq = str(SeqIO.read(fasta_path, format=\"fasta\").seq)\n",
    "\n",
    "m = pd.merge(df, df_true, how=\"outer\", on=[\"pos\", \"wt\"], indicator=True)\n",
    "m = m[m[\"_merge\"] != \"both\"]\n",
    "m[\"codon\"] = m[\"coords\"].apply(lambda x: ref_seq[x[0]: x[1]])\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "4295a99d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt</th>\n",
       "      <th>pos</th>\n",
       "      <th>coords</th>\n",
       "      <th>mut1</th>\n",
       "      <th>num_aa</th>\n",
       "      <th>_merge</th>\n",
       "      <th>codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [wt, pos, coords, mut1, num_aa, _merge, codon]\n",
       "Index: []"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_path = \"/home/yutianc/bjorn_rep/data/sc2/NC_045512.2_escape_BA.4_BA.5_rbd.fa\"\n",
    "csv_path = \"/home/yutianc/sars-cov-2-dms-ref/data/ba.4-ba.5/mut1_BA.4_BA.5_EPI_ISL_11207535.csv\"\n",
    "\n",
    "df = cross_check(fasta_path, region=\"rbd\")\n",
    "df_true = pd.read_csv(csv_path)\n",
    "df_true = df_true.drop_duplicates(subset=[\"pos\"])\n",
    "ref_seq = str(SeqIO.read(fasta_path, format=\"fasta\").seq)\n",
    "\n",
    "m = pd.merge(df, df_true, how=\"outer\", on=[\"pos\", \"wt\"], indicator=True)\n",
    "m = m[m[\"_merge\"] != \"both\"]\n",
    "m[\"codon\"] = m[\"coords\"].apply(lambda x: ref_seq[x[0]: x[1]])\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9988cf19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt</th>\n",
       "      <th>pos</th>\n",
       "      <th>coords</th>\n",
       "      <th>mut1</th>\n",
       "      <th>num_aa</th>\n",
       "      <th>_merge</th>\n",
       "      <th>codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [wt, pos, coords, mut1, num_aa, _merge, codon]\n",
       "Index: []"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_path = \"/home/yutianc/sars-cov-2-dms-ref/output/NC_045512.2_escape_JN.1_rbd.fa\"\n",
    "csv_path = \"/home/yutianc/sars-cov-2-dms-ref/data/jn.1/mut1_JN_1_EPI_ISL_18373905.csv\"\n",
    "\n",
    "df = cross_check(fasta_path, region=\"rbd\")\n",
    "df_true = pd.read_csv(csv_path)\n",
    "df_true = df_true.drop_duplicates(subset=[\"pos\"])\n",
    "ref_seq = str(SeqIO.read(fasta_path, format=\"fasta\").seq)\n",
    "\n",
    "m = pd.merge(df, df_true, how=\"outer\", on=[\"pos\", \"wt\"], indicator=True)\n",
    "m = m[m[\"_merge\"] != \"both\"]\n",
    "m[\"codon\"] = m[\"coords\"].apply(lambda x: ref_seq[x[0]: x[1]])\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "1e4820fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt</th>\n",
       "      <th>pos</th>\n",
       "      <th>coords</th>\n",
       "      <th>mut1</th>\n",
       "      <th>num_aa</th>\n",
       "      <th>_merge</th>\n",
       "      <th>codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [wt, pos, coords, mut1, num_aa, _merge, codon]\n",
       "Index: []"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_path = \"/home/yutianc/sars-cov-2-dms-ref/output/NC_045512.2_escape_KP.2_rbd.fa\"\n",
    "csv_path = \"/home/yutianc/sars-cov-2-dms-ref/data/kp.2/mut1_KP_2_EPI_ISL_18916251_del.csv\"\n",
    "\n",
    "df = cross_check(fasta_path, region=\"rbd\")\n",
    "df_true = pd.read_csv(csv_path)\n",
    "df_true = df_true.drop_duplicates(subset=[\"pos\"])\n",
    "ref_seq = str(SeqIO.read(fasta_path, format=\"fasta\").seq)\n",
    "\n",
    "m = pd.merge(df, df_true, how=\"outer\", on=[\"pos\", \"wt\"], indicator=True)\n",
    "m = m[m[\"_merge\"] != \"both\"]\n",
    "m[\"codon\"] = m[\"coords\"].apply(lambda x: ref_seq[x[0]: x[1]])\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "da382f93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt</th>\n",
       "      <th>pos</th>\n",
       "      <th>coords</th>\n",
       "      <th>mut1</th>\n",
       "      <th>num_aa</th>\n",
       "      <th>_merge</th>\n",
       "      <th>codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [wt, pos, coords, mut1, num_aa, _merge, codon]\n",
       "Index: []"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_path = \"/home/yutianc/sars-cov-2-dms-ref/output/NC_045512.2_escape_KP.3_rbd.fa\"\n",
    "csv_path = \"/home/yutianc/sars-cov-2-dms-ref/data/kp.3/mut1_KP_3_EPI_ISL_19036766.csv\"\n",
    "\n",
    "df = cross_check(fasta_path, region=\"rbd\")\n",
    "df_true = pd.read_csv(csv_path)\n",
    "df_true = df_true.drop_duplicates(subset=[\"pos\"])\n",
    "ref_seq = str(SeqIO.read(fasta_path, format=\"fasta\").seq)\n",
    "\n",
    "m = pd.merge(df, df_true, how=\"outer\", on=[\"pos\", \"wt\"], indicator=True)\n",
    "m = m[m[\"_merge\"] != \"both\"]\n",
    "m[\"codon\"] = m[\"coords\"].apply(lambda x: ref_seq[x[0]: x[1]])\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "02d6d0dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt</th>\n",
       "      <th>pos</th>\n",
       "      <th>coords</th>\n",
       "      <th>mut1</th>\n",
       "      <th>num_aa</th>\n",
       "      <th>_merge</th>\n",
       "      <th>codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [wt, pos, coords, mut1, num_aa, _merge, codon]\n",
       "Index: []"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_path = \"/home/yutianc/sars-cov-2-dms-ref/output/NC_045512.2_escape_XBB.1.5_rbd.fa\"\n",
    "csv_path = \"/home/yutianc/sars-cov-2-dms-ref/data/xbb.1.5/rbd/mut1_XBB_1_5_EPI_ISL_17054053.csv\"\n",
    "\n",
    "df = cross_check(fasta_path, region=\"rbd\")\n",
    "df_true = pd.read_csv(csv_path)\n",
    "df_true = df_true.drop_duplicates(subset=[\"pos\"])\n",
    "ref_seq = str(SeqIO.read(fasta_path, format=\"fasta\").seq)\n",
    "\n",
    "m = pd.merge(df, df_true, how=\"outer\", on=[\"pos\", \"wt\"], indicator=True)\n",
    "m = m[m[\"_merge\"] != \"both\"]\n",
    "m[\"codon\"] = m[\"coords\"].apply(lambda x: ref_seq[x[0]: x[1]])\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2bc7d76",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e13a4603",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c41f05f5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fb4af4f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "379adf15",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "4f55ac92",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_from_dms_ref(path_to_dms_ref, out_name, out_path):\n",
    "    \n",
    "    dms_ref = SeqIO.read(path_to_dms_ref, format=\"genbank\")\n",
    "    (start, end) = next((i.location.start, i.location.end) for i in dms_ref.features if i.type == \"gene\")\n",
    "    gene_seq = dms_ref.seq[start: end]\n",
    "    gene_rec = SeqRecord(gene_seq, id=out_name, name=out_name, description=out_name)\n",
    "\n",
    "    with open(out_path, \"a\") as handle:   # append\n",
    "        SeqIO.write(gene_rec, handle, \"fasta\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "d51bddc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_gene_from_dms_ref(\"/home/yutianc/sars-cov-2-dms-ref/data/ba.1/PacBio_amplicon_BA1.gb\", \"ba.1\", \"/home/yutianc/sars-cov-2-dms-ref/rbd.fa\")\n",
    "get_gene_from_dms_ref(\"/home/yutianc/sars-cov-2-dms-ref/data/ba.2/rbd/PacBio_amplicon_BA2.gb\", \"ba.2\", \"/home/yutianc/sars-cov-2-dms-ref/rbd.fa\")\n",
    "get_gene_from_dms_ref(\"/home/yutianc/sars-cov-2-dms-ref/data/xbb.1.5/rbd/PacBio_amplicon.gb\", \"xbb.1.5\", \"/home/yutianc/sars-cov-2-dms-ref/rbd.fa\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "9f8d643e",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_gene_from_dms_ref(\"/home/yutianc/sars-cov-2-dms-ref/data/ba.2/spike/PacBio_amplicon.gb\", \"ba.2\", \"/home/yutianc/sars-cov-2-dms-ref/spike.fa\")\n",
    "get_gene_from_dms_ref(\"/home/yutianc/sars-cov-2-dms-ref/data/xbb.1.5/spike/PacBio_amplicon.gb\", \"xbb.1.5\", \"/home/yutianc/sars-cov-2-dms-ref/spike.fa\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e497052d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02077ac4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "FA_EXTS = {\".fa\", \".fasta\", \".fna\"}\n",
    "\n",
    "def iter_fasta_files(root: str):\n",
    "    root_path = Path(root)\n",
    "    for p in root_path.rglob(\"*\"):\n",
    "        if p.is_file() and p.suffix.lower() in FA_EXTS:\n",
    "            yield p\n",
    "\n",
    "\n",
    "def extract_ids_from_fasta_file(fp: Path):\n",
    "    ids = []\n",
    "    fname = []\n",
    "    with fp.open(\"r\", encoding=\"utf-8\", errors=\"replace\") as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\">\"):\n",
    "                # first token after '>' (drops description)\n",
    "                seq_id = line[1:].strip().split()[0]\n",
    "                if seq_id:\n",
    "                    try:\n",
    "                        sra = seq_id.split('/')[2]\n",
    "                        fa_name = (f.name.split(\"/\")[-1]).split('.')[0]\n",
    "                        if sra != fa_name:\n",
    "                            ids.append(sra)\n",
    "                            fname.append(fa_name)\n",
    "                    except:\n",
    "                        continue\n",
    "    return ids, fname\n",
    "\n",
    "def extract_ids_from_dir(root: str):\n",
    "    out_id, out_fname = [], []\n",
    "    for fp in iter_fasta_files(root):\n",
    "        ids, fname = extract_ids_from_fasta_file(fp)\n",
    "        out_id.extend(ids)\n",
    "        out_fname.extend(fname)\n",
    "    print(len(out_fname))\n",
    "    print(len(out_id))\n",
    "\n",
    "    return pd.DataFrame({\"ID\": out_id, \"fname\": out_fname})\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    root_dir = \"/home/yutianc/bjorn_rep/data/sc2/consensus_sequences/\"\n",
    "    df = extract_ids_from_dir(root_dir)\n",
    "\n",
    "d = dict(zip(df[\"fname\"], df[\"ID\"]))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
